Tests/MongoSwiftSyncTests/SpecTestRunner/TestOperations/TestOperation.swift

```diff
@@ -0,0 +1,298 @@
+
+    func execute(on database: MongoDatabase, sessions: [String: ClientSession]) throws -> TestOperationResult?
+
+    func execute(
+        on collection: MongoCollection<Document>,
+        sessions: [String: ClientSession]
+    ) throws -> TestOperationResult?
+
+    func execute(on session: ClientSession) throws -> TestOperationResult?
+
+    func execute<T: SpecTest>(on runner: inout T, sessions: [String: ClientSession]) throws -> TestOperationResult?
+
+    /// The name of the session this operation should execute against, if any.
+    var session: String? { get }
+}
+
+extension TestOperation {
+    func execute(on _: MongoClient, sessions _: [String: ClientSession]) throws -> TestOperationResult? {
+        throw TestError(message: "\(type(of: self)) cannot execute on a client")
+    }
+
+    func execute(on _: MongoDatabase, sessions _: [String: ClientSession]) throws -> TestOperationResult? {
+        throw TestError(message: "\(type(of: self)) cannot execute on a database")
+    }
+
+    func execute(
+        on _: MongoCollection<Document>,
+        sessions _: [String: ClientSession]
+    ) throws -> TestOperationResult? {
+        throw TestError(message: "\(type(of: self)) cannot execute on a collection")
+    }
+
+    func execute(on _: ClientSession) throws -> TestOperationResult? {
+        throw TestError(message: "\(type(of: self)) cannot execute on a session")
+    }
+
+    func execute<T: SpecTest>(on _: inout T, sessions _: [String: ClientSession]) throws -> TestOperationResult? {
+        throw TestError(message: "\(type(of: self)) cannot execute on a test runner")
+    }
+
+    var session: String? { nil }
+
+    func getSession(from sessions: [String: ClientSession]) throws -> ClientSession? {
+        guard let sessionName = self.session else {
+            return nil
+        }
+
+        guard let session = sessions[sessionName] else {
+            throw TestError(message: "\(sessionName) not included in sessions map \(sessions)")
+        }
+
+        return session
+    }
+}
+
+/// A enumeration of the different objects a `TestOperation` may be performed against.
+enum TestOperationObject: Decodable {
+    case client
+    case database
+    case collection
+    case gridfsbucket
+    case testRunner
+    case session(String)
+
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.singleValueContainer()
+        let rawValue = try container.decode(String.self)
+        switch rawValue {
+        case "client":
+            self = .client
+        case "database":
+            self = .database
+        case "collection":
+            self = .collection
+        case "gridfsbucket":
+            self = .gridfsbucket
+        case "testRunner":
+            self = .testRunner
+        default:
+            self = .session(rawValue)
+        }
+    }
+}
+
+/// Struct containing an operation and an expected outcome.
+struct TestOperationDescription: Decodable {
```

Comment:
the rest of the stuff in the file is pretty much unchanged besides adding new decoding logic for new operations.